_examples/dupsteps/cmd/stepchecker/main.go

```diff
@@ -0,0 +1,295 @@
+package main
+
+import (
+	"bufio"
+	"fmt"
+	"go/ast"
+	"go/parser"
+	"go/token"
+	"log"
+	"os"
+	"regexp"
+	"strconv"
+	"strings"
+)
+
+//
+// See accompanying README file(s);
+// also https://github.com/cucumber/godog/pull/642
+//
+
+func main() {
+	if len(os.Args) < 3 {
+		log.Printf("Usage: main.go [go-file(s)] [feature-file(s)]\n")
+
+		os.Exit(RC_USER)
+	}
+
+	// Structures into which to collect step patterns found in Go and feature files
+	godogSteps := make(map[string]*StepMatch)
+	featureSteps := make(map[string]*StepMatch)
+
+	// collect input files (must have at least one of each kind (e.g., *.go, *.feature)
+	for _, filePath := range os.Args[1:] {
+		if strings.HasSuffix(filePath, ".go") {
+			if err := collectGoSteps(filePath, godogSteps); err != nil {
+				fmt.Printf("error collecting `go` steps: %s\n", err)
+
+				os.Exit(RC_ISSUES)
+			}
+		}
+
+		if strings.HasSuffix(filePath, ".feature") {
+			if err := collectFeatureSteps(filePath, featureSteps); err != nil {
+				fmt.Printf("error collecting `feature` steps: %s\n", err)
+
+				os.Exit(RC_ISSUES)
+			}
+		}
+	}
+
+	if len(godogSteps) == 0 {
+		log.Printf("no godog step definition(s) found")
+
+		os.Exit(RC_USER)
+	}
+
+	if len(featureSteps) == 0 {
+		log.Printf("no feature step invocation(s) found")
+
+		os.Exit(RC_USER)
+	}
+
+	// Match steps between Go and feature files
+	matchSteps(godogSteps, featureSteps)
+
+	var issuesFound int
+
+	// Report on unexpected (i.e., lack of, duplicate or ambiguous) mapping from feature steps to go steps
+	fmt.Printf("Found %d feature step(s):\n", len(featureSteps))
+
+	var fsIdx int
+
+	for text, step := range featureSteps {
+		fsIdx++
+
+		fmt.Printf("%d. %q\n", fsIdx, text)
+
+		if len(step.matchedWith) != 1 {
+			issuesFound++
+
+			fmt.Printf("  - %d matching godog step(s) found:\n", len(step.matchedWith))
+
+			for _, match := range step.source {
+				fmt.Printf("    from: %s\n", match)
+			}
+
+			for _, match := range step.matchedWith {
+				fmt.Printf("      to: %s\n", match)
+			}
+		}
+	}
+
+	fmt.Println()
+
+	// Report on lack of mapping from go steps to feature steps
+	fmt.Printf("Found %d godog step(s):\n", len(godogSteps))
+
+	var gdsIdx int
+
+	for text, step := range godogSteps {
+		gdsIdx++
+
+		fmt.Printf("%d. %q\n", gdsIdx, text)
+
+		if len(step.matchedWith) == 0 {
+			issuesFound++
+
+			fmt.Printf("  - No matching feature step(s) found:\n")
+
+			for _, match := range step.source {
+				fmt.Printf("    from: %s\n", match)
+			}
+		}
+	}
+
+	fmt.Println()
+
+	if issuesFound != 0 {
+		log.Printf("%d issue(s) found\n", issuesFound)
+		os.Exit(RC_ISSUES)
+	}
+}
+
+func collectGoSteps(goFilePath string, steps map[string]*StepMatch) error {
```

Comment:
I dont know if this static parser will spot the cases where the ScenarioContext has been wrapped.
I've seen (and I use) a scheme where folk register steps via a proxy wrapper of the ScenarioContext, that adds generic behaviours to the steps being registered.

Reply:
Nice catch! 👍

If you provide an example, I'll work up a test case to fix.

Do you have a fix strategy in mind?